### PR TITLE
Remote-Hosting of Certain Package Data

### PR DESCRIPTION
**Pull request by tbowers7**
The main thrust of this PR is to reduce the size of the PyPI-hosted release packages by omitting various package-included data such as `reid_arxiv` and `sensfunc` files.  

Major Changes:
* Add download/caching for various package-included data using AstroPy's [`download_file()`](https://docs.astropy.org/en/stable/api/astropy.utils.data.download_file.html#astropy.utils.data.download_file) functionality.  Remove data files from package export in the `MANIFEST.in` file.
* Use this framework to directly pull Telluric Atmospheric Grid files from the cloud into the user's cache without separate download from Google Drive.
* Also use this framework to save the output of `pypeit_identify` (`wvarxiv.fits`) into the user's cache as a way to directly and quickly use manual wavelength identifications in reductions.
* Add unit tests to ensure the downloading introduced in this PR continues to work.

Minor Changes:
* Tweak the `pypeit_identify` GUI to display the spectrum on log scale by default (linear scale selectable by command line argument).
* Address #1345, since I was tweaking wavelength stuff anyways.

Further Discussion:
Especially since most users will not need most of the files in the `pypeit.data` directories (and they are somewhat large), this PR adds functionality to download them from the GitHub repo at runtime when needed.

The downloading and caching process takes advantage of AstroPy's [`download_file()`](https://docs.astropy.org/en/stable/api/astropy.utils.data.download_file.html#astropy.utils.data.download_file) functionality, along with user-space caching.  Downloaded files are nominally kept in `~/.pypeit/cache`, but this can be changed using the [Astropy configuration system](https://docs.astropy.org/en/stable/config/index.html).  For files being kept out of the release distribution, PypeIt now checks first the package directory, then the cache, then downloads the file if not already found.

Taking advantage of this new framework, this PR also revamps the `pypeit_install_telluric` script to download a specified Telluric Atmospheric Grid file from the cloud into the cache for use.  This alleviates the need for separately downloading said files from the Google Drive and installing them into the package directory.

A feature requested by `LDT/DeVeny` users is the ability to quickly and easily use manual wavelength calibrations for their reductions.  This PR uses the new caching functionality to save a copy of the output `wvarxiv.fits` from `pypeit_identify` (with date/time in the filename) for direct reuse with the `full_template` method.

Unit Tests Pass:
```
============================= test session starts ==============================
platform linux -- Python 3.9.12, pytest-7.1.1, pluggy-1.0.0
3.7.0, doctestplus-0.12.0, hypothesis-6.41.0
collected 329 items                                                            
======================= 329 passed in 1868.46s (0:31:08) =======================
```

Dev Suite (branch: `develop`) is getting caught on the newer `keck_lris_red_mark4` files -- also a `gemini_gnirs` error "No matching objects for OBJ0659-DET01", but otherwise passes.  May need to run the `dev-suite` in the cloud.

### CODE REVIEW

**Comment by tbowers7** [comment]
The CI tests are revealing some issues with pathnames and filenames in this new scheme.  I will need to address them before this PR can be reviewed.

File: pypeit/core/gui/identify.py
Comment (profxj) [inline pypeit/core/gui/identify.py:283]: Nice added feature

File: pypeit/par/pypeitpar.py
Comment (profxj) [inline pypeit/par/pypeitpar.py:2490]: well pruned

File: pypeit/spectrographs/gtc_osiris.py
Comment (profxj) [inline pypeit/spectrographs/gtc_osiris.py:279]: @rcooke-ast -- you might confirm this is ok

File: pypeit/spectrographs/wht_isis.py
Comment (profxj) [inline pypeit/spectrographs/wht_isis.py]: also a quick check for @rcooke-ast

File: pypeit/tests/test_pkgdata.py
Comment (profxj) [inline pypeit/tests/test_pkgdata.py:1]: Thanks for the tests!!

**Review by profxj — APPROVED** [review]
Great stuff.

Should we consider a test in the DevSuite too?
Or do you feel what you've generated is enough?

Am approving either way.  But maybe wait for @rcooke-ast to comment on the elmination of semi-brute and basic.  Thanks!

File: pypeit/core/gui/identify.py
Comment (rcooke-ast) [inline pypeit/core/gui/identify.py:289]: Perhaps be explicit, and add:
`ax.set_yscale('linear')`

I can see reasons why we might want to toggle between log and linear during the fitting (e.g. the user... likely me... forgets to put the command line flag)

File: pypeit/core/gui/identify.py
Comment (rcooke-ast) [inline pypeit/core/gui/identify.py]: Should we add the explicit lines of code the user needs to copy paste into their reduction file?

```
[calibrations]
  [[wavelengths]]
    reid_arxiv = {cachename}
    method = full_template
```

File: pypeit/par/pypeitpar.py
Comment (rcooke-ast) [inline pypeit/par/pypeitpar.py:2553]: Yes! Be gone `basic` and `semi-brute`!  I don't think they are even used...

File: pypeit/spectrographs/wht_isis.py
Comment (rcooke-ast) [reply pypeit/spectrographs/wht_isis.py]: Yes, this can be deleted

File: pypeit/spectrographs/gtc_osiris.py
Comment (rcooke-ast) [reply pypeit/spectrographs/gtc_osiris.py:279]: Yes, fine to comment it out, but maybe don't delete just yet - thanks!

**Comment by tbowers7** [comment]
> Should we consider a test in the DevSuite too?
> Or do you feel what you've generated is enough?

@profxj : I think an independent `dev-suite` run in the cloud will be very useful given the new feature of needing to download certain data from a remote server.  I found several issues just with the CI tests, and I would appreciate having a separate `dev-suite` run not on my own machine.

File: pypeit/par/pypeitpar.py
Comment (tbowers7) [reply pypeit/par/pypeitpar.py:2553]: They are not used... I did a thorough check of all non-deprecated code and found they are commented out.  I chose not to delete the commented code sections entirely, but maybe we can consider moving them to `deprecated/` somehow and remove the commented code from `pypeit/core/arc.py`.

**Comment by tbowers7** [comment]
@tepickering : I'm interested in your view of how I implemented AstroPy's [`download_file()`](https://docs.astropy.org/en/stable/api/astropy.utils.data.download_file.html#astropy.utils.data.download_file) functionality, and if it conforms to standard usage.

File: pypeit/core/gui/identify.py
Comment (tbowers7) [reply pypeit/core/gui/identify.py]: Added.

File: pypeit/core/gui/identify.py
Comment (tbowers7) [reply pypeit/core/gui/identify.py:289]: Added.

File: pypeit/data/utils.py
Comment (tepickering) [inline pypeit/data/utils.py]: small typo

**Review by tepickering — APPROVED** [review]
really great stuff! i heartily approve!

**Comment by tepickering** [comment]
> @tepickering : I'm interested in your view of how I implemented AstroPy's [`download_file()`](https://docs.astropy.org/en/stable/api/astropy.utils.data.download_file.html#astropy.utils.data.download_file) functionality, and if it conforms to standard usage.

looks good to me. like i said in the slack convo, ours is the exact use case that it was built for and you're using it exactly as intended.

**Comment by tbowers7** [comment]
Cloud run of `dev-suite` passes with only 1 (well-documented) error:
```
And it is our (not so favorite):
End of Log:
  File "/PypeIt/pypeit/pypmsgs.py", line 185, in error
    raise PypeItError(msg)
pypeit.pypmsgs.PypeItError: No matching objects for OBJ0211-DET01.  Odds are you input the wrong OBJID
```

I will add a little more documentation about how to run PypeIt without internet access (_i.e._, how to pre-download necessary files) before merging.

**Comment by tbowers7** [comment]
@kbwestfall : I would appreciate your eye on these last few changes, in addition to any comments you have on the PR overall.

Some more doc changes forthcoming.

File: doc/installing.rst
Comment (kbwestfall) [inline doc/installing.rst:279]: Awesome!

File: doc/wave_calib.rst
Comment (kbwestfall) [inline doc/wave_calib.rst]: Should this be `wvarxiv.fits` to match the example in the text?

File: pypeit/data/utils.py
Comment (kbwestfall) [inline pypeit/data/utils.py]: You don't need the function name in the docstring.

File: pypeit/data/utils.py
Comment (kbwestfall) [inline pypeit/data/utils.py]: I don't think the above will render correctly.  Check the syntax and distinction between Google and Numpy styles: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/#google-vs-numpy .  You can pick either, but the syntax has to consistently conform to one of the two within any given docstring.

File: pypeit/sensfilearchive.py
Comment (kbwestfall) [inline pypeit/sensfilearchive.py]: Can we change this to be explicitly `copy_to_pkgdir`?  I definitely appreciate the intention to test this with a unit test.  But, as much as possible, functions should not "know" whether or not they're being executed as part of a unit test.

**Review by kbwestfall — CHANGES_REQUESTED** [review]
Amazing work, @tbowers7 !!  Thanks for all your effort on this.  I have a couple minor requests, though:

 - Please check the docstring syntax.  I don't think Sphinx will correctly render what you have in many of the `data.utils` functions.
 - I realize I'm in the minority, but the use of `as` is a pet peeve of mine.  Unless it's an "industry standard" (like `import numpy as np`), I think we should almost always avoid using this.  Specifically, instead of `from astropy.utils import data as astropy_data`, I'd prefer `import astropy.utils.data`; and instead of `from pypeit import __version__ as pypeit_version`, I'd prefer `from pypeit import __version__`.  And then propagate the relevant changes.

Three other thoughts/questions not necessarily for this PR:

- Similar to your excellent changes to the telluric installation script, is it useful to update the QL masters installation script?
- Can you remind me of the use case for the `copy_to_pkgdir` option?  Is it possible/useful to instead have a symlink installed in the package directory?  I'm wary of having multiple files on disk, particularly for the large telluric files.
- For the telluric files, can we eventually get all the telluric files in the cloud?  I think I remember that being the goal, but I don't remember the hang-up.

**Comment by profxj** [comment]
Regarding @kbwestfall 's comments on the QL Masters' files, 
  I am going to generate Cloud scripts for `release` and `develop`
 that will run `build_ql_masters` and push the outputs to `s3`.  
 And expose them.

I'll then ask @tbowers7 for help to integrate them into the 
  framework he build for this PR.  Bottom line, defer the masters files
  until a (not so) future PR.  Thx!

File: doc/wave_calib.rst
Comment (tbowers7) [reply doc/wave_calib.rst]: I've addressed this in 2 ways.
  1. No longer writing to generic `wvarxiv.fits`, but instead writing to `wvarxiv_{spec}_{datetime}.fits` in the local directory.
  2. I've corrected the text above this code block to indicate what is saved in the cache is labeled with the spectrograph and timestamp of creation for uniqueness.  This is in case a user is doing manual IDs for multiple setups and wants to be able to use all of them.

**Comment by tbowers7** [comment]
@kbwestfall :
> * Can you remind me of the use case for the `copy_to_pkgdir` option?  Is it possible/useful to instead have a symlink installed in the package directory?  I'm wary of having multiple files on disk, particularly for the large telluric files.

This option is only necessary for a single unit test case to appease an `assert` statement regarding the name of a `sensfunc` file.  I've removed the option from the `reid_arxiv` and `skisim` functions, and have converted the remaining instance to a symlink situation.

> * For the telluric files, can we eventually get all the telluric files in the cloud?  I think I remember that being the goal, but I don't remember the hang-up.

@profxj has uploaded to the cloud the telluric files specifically called out in spectrograph parameter lists.  I'll defer to him the question of putting the remaining files there.

**Comment by profxj** [comment]
There is no problem putting all the `telluric` files in the cloud.
  I simply prioritized the ones required for redux

**Comment by kbwestfall** [comment]
Sorry to harp on this, @tbowers7 , but the docstring syntax still won't render correctly.  Here's an example for the `get_reid_arxiv_filepath` function:

Current:
```
    Args:
        arxiv_file: str
          The base filename of the ``reid_arxiv`` file to be located

    Returns:
        calibfile: str
          The full path to the ``reid_arxiv`` file
```
Renders as:
incorrect_rendering

Corrected Google style:
```
    Args:
        arxiv_file (str):
            The base filename of the ``reid_arxiv`` file to be located

    Returns:
        str: The full path to the ``reid_arxiv`` file
```
Renders as:
correct_google_rendering

Corrected Numpy Style:
```
    Parameters
    ----------
    arxiv_file: str
        The base filename of the ``reid_arxiv`` file to be located

    Returns
    -------
    calibfile: str
        The full path to the ``reid_arxiv`` file
```
correct_numpy_rendering

Everything else looks great!

**Comment by tbowers7** [comment]
@kbwestfall : Thanks for your patience with the docstring syntax.  Sphinx is still on my "to learn" list.

**Comment by kbwestfall** [comment]
Thanks for being willing to bear with me...  There's one final issue, which is a subtle difference between the return syntax for Google and numpy styles.  Google style doesn't allow you define the name of the returned object, whereas you can with numpy format; i.e., the fact that I did not include the name of the returned object in the corrected syntax above is important for the former.  You're only allowed to define the type and provide a description in Google style.  This means that when you're returning multiple objects, the return type is always `tuple`, and you provide the components of the tuple in the description.  If you think it's important to provide the object names in the return statement, you have to use Numpy style.

**Comment by tbowers7** [comment]
If it's still not right, maybe you can expose this branch on readthedocs so I can continue tweaking until it's right.

I usually use Numpy style (but not with Sphinx) for my docstrings, but was trying to match existing PypeIt style.

**Comment by kbwestfall** [comment]
The quick way to check how things render is to make the html locally:
```
cd doc
make clean ; make html
open _build/html/index.html
```
and then navigate to the relevant page.  All the function docs are accessed through the `Python API` link in the navigation bar.  The `make html` command also runs a bunch of scripts to create some of the auto-documented files (like the datamodel and parameters).  If you don't need to remake these (i.e., if you're just editing function doc strings), you can speed up the process by just calling `make htmlonly`, instead of `make clean ; make html`.  This still takes a while, but it should be faster than waiting for an exposed branch on readthedocs to rebuild.

The "PypeIt style" is a bit all over the place.  I prefer Google style, Joe prefers Numpy style, and there are still many functions with syntax errors.  If you're more comfortable with Numpy style, I would stick with it and just make sure that you use syntax that Sphinx can correctly render in html.

I'll go ahead and approve now, but it looks like there are some CI failures that should be fixed before merging.

Thanks again for all your effort on this!

**Review by kbwestfall — APPROVED** [review]

**Pull request merged**
